### DISCUSSION TITLE
MAC DMG

### DISCUSSION THREAD

**Discussion by gitbass99** [RomanKubiak/ctrlr#521] [Releases]
Guys, where are the latest DMGs? I don't know anything about coding.

**Comment by dnaldoog** [comment]
Download this one :)

https://ctrlr.org/nightly/Ctrlr-5.3.198.dmg

**Reply by gitbass99** [reply]
Thank you!

**Reply by now-its-dark** [reply]
It seems that version one performs very poorly on an ARM-based mac— doing anything causes the application to hang for a time and the framerate when in panel mode is very choppy. Do you happen to know if there is a version which is stable and runs well on ARM macs?

**Reply by damiensellier** [reply]
There's a macOS M1 version for Ctrlr 5.6 (commit f452347) build available , I downloaded it when I saw it but I don't have a M1 or M2 so I'm not sure if it's working well:

Ctrlr 5.6 Mac ARM Standalone: 
[Ctrlr-5.6-f452347-App.zip](https://github.com/RomanKubiak/ctrlr/files/10947199/Ctrlr-5.6-f452347-App.zip)

Ctrlr 5.6 Mac ARM Appex: 
[Ctrlr-5.6-f452347-appex.zip](https://github.com/RomanKubiak/ctrlr/files/10947200/Ctrlr-5.6-f452347-appex.zip)

Ctrlr 5.6 Mac ARM AU: 
[Ctrlr-5.6-f452347-component.zip](https://github.com/RomanKubiak/ctrlr/files/10947201/Ctrlr-5.6-f452347-component.zip)

Ctrlr 5.6 Mac ARM VST2: 
[Ctrlr-5.6-f452347-vst.zip](https://github.com/RomanKubiak/ctrlr/files/10947202/Ctrlr-5.6-f452347-vst.zip)

Ctrlr 5.6 Mac ARM VST3: 
[Ctrlr-5.6-f452347-vst3.zip](https://github.com/RomanKubiak/ctrlr/files/10947203/Ctrlr-5.6-f452347-vst3.zip)

Let us know if it's working properly. Thanks
Damien

**Reply by now-its-dark** [reply]
Hi @damiensellier, thanks for sharing these links. This is the version I've been using, but my understanding was that 5.6 has a number of issues, it seems like 5.3.* is often the suggested stable version for use in general, is this no longer the case? 

I've noticed that some of the official demo panels don't run correctly with this version, such as the "Render" example.

**Comment by maxtr0ne** [comment]
anyone ever get their finger on the pulse of where a stable version(s) for Mac M1+ are located? i'm also not a coder but fairly tech savvy

**Reply by damiensellier** [reply]
Unfortunately there's no such thing as a "stable" version of ctrlr. v5.3.198 is working great but only available as vst2.4 and standalone for mac intel linux and windows.
v5.6.1 has its quirks because of some juce deprecated codes but I prefer working with it in the end because it can deliver VST3 plugins. It's available for mac pc as standalone vst2 vst3 AU and can be built for M1 from xcode. So if you want to dive deeper to provide a perfect  M1 version that would be awesome. I still don't have a M1 mac so not able to take a look at it.

**Reply by maxtr0ne** [reply]
longtime user on an older mac, and now i have a m1 & i'd love to try, but i don't know how to take these files above and make an executable installer or whatever it's called. if i try to drop the files into the plugin folders, the Mac rejects them when the plugin scan runs in ableton live & logic pro, and the mac suggests you throw them in the trash due to unknown developer

**Reply by damiensellier** [reply]
There's a simple procedure to skip the macos gatekeeper and add file's attributes to let the VST running on newer mac. 

https://youtu.be/S-FH9Tlkj5M?t=222

Never tried since I'm on an intel mac but you should give it a try.

**Reply by robucsc** [reply]
Hey Damien,

I've been using the 5.6 version you posted on an M2 with Ventura the last few days and so far it's doing fine. I'm still in just the panel building phase though, and haven't done anything extensive. Just learning how to make panels. Thanks for posting it as I was having trouble loading images with 5.3.

**Reply by damiensellier** [reply]
@robucsc Awesome! Let us know if you need help with your panel, you should be able to export Standalone macOS apps from the Ctrlr APP, VST2 pannels from the CTRLR VST2 plugin in your DAW and AU plugins as well without any problems. 
The VST3 is currently giving identification issues though, and requires building a VST3 for each panel you want to export with the Projucer and XCode with different name, IDs, plugin manufacturer's name etc to identify them properly.

**Reply by robucsc** [reply]
@damiensellier Hi Damien,

That's very kind of you. It's mostly going well. I'm just doing CC stuff so far. But I have run into something I've not yet been able to figure out. One of the parms is a list, and I know how to make a fixed slider, but not set the value for each state other than the default. So the first is 0 then 2, 16,  32, 48, 60, 127.

Once I get this figure out I'll likely start trying to make an AU version next. One step at at time. ;) 

I do have the Projucer and Xcode on the machine already.

**Reply by robucsc** [reply]
On this docs [page](https://ctrlr.org/expressions-in-ctrlr-2/) near the bottom Chab asks for a map, which is the same thing as a hash table. That seems like it would be the way to go. Looks like it's time to learn to Lua. But also for now I can do something ugly like:
eq(modulatorValue, 0, 0, eq(modulatorValue, 1, 2, eq(modulatorValue, 3, 16, eq(modulatorValue, 4, 32, eq(modulatorValue, 5, 48, eq(modulatorValue, 6, 60, eq(modulatorValue, 7, 127)))))))

And the reverse for the other direction. ;)

**Reply by robucsc** [reply]
eq(value, 0, 0, eq(value, 2, 1, eq(value, 16, 3, eq(value, 32, 4, eq(value, 48, 5, eq(value, 60, 6, eq(value, 127, 7)))))))

**Reply by robucsc** [reply]
I've made more progress, and the panel is basically doing all that I had set out to achieve with this first version of it. The next step is to compile it into an AU plugin. I have the Projucer, and Xcode. When I do _save and open_ it gets to Xcode and fails with a 
    'boost/config.hpp' file not found
error. 

It seems boost is a C++ library, so I did a 
    brew install boost
Which installed boost, but upon trying to compile again I got the same error. I'm hunting around for solutions, and it seems a fair number of other projects that employ boost run into this same issue. I'm on the hunt, but I've yet to find a solution.

**Reply by damiensellier** [reply]
Hi @robucsc,
 No need to build anything with Projuicer/Xcode to get AU plugins. 
To create AU plugins, you just need to put the ctrlr.component I posted above into your library/Audio/plugins/Component folder. Maybe you'll have to follow the procedure to bypass the gatekeeper for unsigned applications also posted above.
https://youtu.be/S-FH9Tlkj5M?t=222

Load the Ctrlr plugin in Logic, open your panel and do File>Export> Export instance or Export as restricted instance...

Concerning Boost, if you need to build Ctrlr, Boost.zip is already included in the Ctrlr binaries BUT it needs to be unzipped within this specific path ../Ctrlr/Source/boost/boost/(all the boost files here)

For more info about building CTRL : https://godlike.com.au/fileadmin/godlike/techtools/ctrlr/guides/My_guide_to_compiling_Ctrlr_for_macOS__Mojave__v2.pdf

Good luck!

**Reply by robucsc** [reply]
Hi @damiensellier ,

Thanks for explaining how to make an instance. That's actually what I wanted to do. And thanks for the YouTube link, usually I do this through the Privacy Security control panel, but it's nice to be able to do it from the command line too. 

I did try to get Ctrlr to compile. I moved things around like you suggested, and like in the pdf you linked. I was able to get past the first boost issue, but then ran into some other issues, which I should probably save for a new thread.

I was able to load the Ctrlr stuff in Logic, but I don't use Logic. I only bought it to be able to work with people that do. I generally used Digital Performer. I also tried it in Reaper. Only Logic seems to see it so far, and only as a midi plug in. I've been using the Matrix 6 Ctrlr panel that [Possemo](https://ctrlr.org/author/possemo/) made. That shows up properly in DP, as a plugin instrument. Does the same in Logic too.

It looks like I can make a stand alone version, which should be enough for now.

Thanks!

**Reply by damiensellier** [reply]
Glad to help you. Maybe you could see the Matrix 6 Panel in DP because it's an intel version running in compatibility mode.
I never saw a M1 ctrlr panel official release yet, you'll be the first one.
I invite you to create a new thread dedicated to your building problem in Xcode. In projucer you can select which type of plugin and category you want your DAW to sort it. You'll probably need to check "requires midi" also "is an instrument" and something like "tool" or "Synth" if the panel is for a synth as the sub category for the AU definitions.
Good luck with your project.

**Reply by robucsc** [reply]
Where to begin. 

I can make a stand alone instance, but it doesn't communicate via midi. I made one earlier when I wasn't connected to the synth. Now that I'm connected there doesn't seem to be any communication, even if I make an instance while the synth is connected. There is still communication from the main app, but I'd like to be able to get it to be some sort of independent sort of panel., like what Possemo was able to achieve. His work is actually what inspired me to try to use Ctrlr.

When doing a compile is that just for the main app, or does it also have the ability to produce different sorts of plugins/apps?

Also, with in the main app I can send midi, and I can see it return from the synth with the midi monitor, but it's not coming back and updating the CC knobs.

Ha! It's fairly ambitious to think that I'll have an official release. I would like to share this with folks though, that would be nice. :) Right now I'm just trying to make a CC panel that covers some hidden features of the synth that cannot be edited from the synth. Extra voice modes, filter modes, etc. Pretty fun stuff. If I can achieve that I'll be pretty happy with it.

I do appreciate the support. :)

**Reply by damiensellier** [reply]
On your exported instance did you select again your midi input/output device properly from the top menu?
![MIDI DEVICES I/O](https://github.com/RomanKubiak/ctrlr/assets/55164385/55aee216-a2b2-461b-a639-dc0f364d34bb)

If the midi was working from the panel loaded in the stock Ctrlr AU, it should work from your exported instance (it's a clone, with your panel files embeded), but it probably requires you set up your midi i/o settings again. 
If the midi is not working on the M1 version of Ctrlr, there's probably an issue with Ctrlr itself in M1 mode. Then I suggest you use the old CTRLR for Intel in compatibilty mode.

About compiling CTRLR, (not sure if you really need that since you already have the different M1 builds available), you first have to define parameters in the Projucer Project settings window:
![Projucer Project Settings Window](https://github.com/RomanKubiak/ctrlr/assets/55164385/70812514-6905-4c01-a56f-6246c5f2c5b7)

This is where you should be able to change a few things from the original Ctrlr.jucer file if things don't work properly from the get go.
You can tell Projucer the name of the built plugins/app, the name of the manufacturer that your DAW is going to see, as well as plugin categories, if it's a virtual synth (required for VST2 to get midi in/out messages) etc

You also have in Projucer the Exporter tabs which defines the details for the different IDE depending on the platform. For Xcode you need to check the macOS Architecture option to check if it's relevant to the M1/M2 setup.
When you're done, you save your .jucer project and click on the white button on top right of the project window to open the project in Xcode.
![IDE Settings](https://github.com/RomanKubiak/ctrlr/assets/55164385/897ecdee-2c0a-4faf-a9fb-f45afb274c19)


In Xcode, follow all the details from the PDF tutorial (link above). You'll be able to chose your output type here:
![Xcode output Type](https://github.com/RomanKubiak/ctrlr/assets/55164385/1db90b15-de18-4f03-91fe-b6ca7d7a231a)

You can either chose a standalone app (release mode), VST2, VST3, AU, AUv3 etc 
Follow the different steps from the pdf documentation and first build "Shared Code" in "release mode"

**Reply by robucsc** [reply]
Hmmm.

First thanks for such an awesome post.

I've been making an instance from the Ctrlr app itself, and not from the AU plug in (since that didn't work so well). What I get is an app that opens and acts like it is working, but no midi to the synth.

The instance version doesn't have any menus, and no way that I can see to set the midi, like Poss's M6 panel. He's doing in on one of his tabs, but I'm only doing it with the menu items within the Ctrlr app.

Also, from within the Ctrlr app I can see midi coming in from the synth via the midi monitor, but it doesn't talk to the panel. I've not found a setting that tells the panel to listen to external midi. I have set up the Midi Controller, Input, and Output to the synth, and I can see the midi from the synth in the midi monitor window.

I was somehow able to get the AU plugin to load into DP—this happened when I added another plugin to my system today, but when I try to access it the slider window shows up rather than the knobs from the panel, and then it crashes.

I haven't tried compiling anything today.

The reason I went to 5.6 is because 5.3 would not load images into the knobs on my machine. I haven't tried loading panels made with 5.6 into 5.3, but I suspect that they wouldn't load.

**Reply by damiensellier** [reply]
When you export an instance you can either chose a regular instance or a restricted instance. 
Restricted instance means you won't be able to edit the panel and some of the options in the menu bar will be greyed out. 

On the edit panel, when you design it, in the global properties "Editor" you'll see an option "Is menu bar visible for this panel", and another one "Hide the menu bar when exporting an instance".  You can also decide to show the menu bar but hide the midi menu with the options below. So I invite you to check if everything is set properly to your need before exporting your instance. 

<img width="450" alt="Capture d’écran 2023-07-17 à 01 26 41" src="https://github.com/RomanKubiak/ctrlr/assets/55164385/ee67b523-a942-4609-858b-ba2febbe3aa3">

Please also check your panel has a unique name and you set a "panel author name". It's important to identify your plugin properly in your DAW. It's also important to increment your panel versions in the appropriate box.

Don't hesitate to share some screen captures of your different settings and the results you have, it will help a lot.

The uiImageSlider drop down menu not showing the image resource list is the main reason I also switched to 5.6 from 5.3. It's a bummer when designing the panels.

I recommand starting using the uiSliders first to see if the communication is working properly and evaluate the different settings required. Once everything is fine you can switch to a custom image slider with the same settings for a more appealing design.  It's always good to have a scratch panel to make test before diving to deep into the GUI part.

Did you set everything properly in your different sliders props?
<img width="449" alt="slider properties" src="https://github.com/RomanKubiak/ctrlr/assets/55164385/10d21df0-d9c0-4099-b637-bb7022228c21">


Damien

**Reply by robucsc** [reply]
It looks like I've had the proper setting selected the entire time. Still the menu doesn't show up when I make an instance with 5.6.

I opened up the panel in 5.3. All the work I had done in 5.3 was there, including the knobs from knobman. I then did an export. Tested the export. The menus showed up, and I was able to select the proper midi setup, and it works. This exports an Intel version. 

The 5.6 version exports a Universal version. The Intel version is slightly larger than the universal JUCE version. In both cases I did stand alone, not plug in versions. I have done plugins with the JUCE version, but not today.

It would seem that with the current setup (5.6) isn't able to make ARM versions. But there is a work around for now, by editing in 5.6, and then using 5.3 to make Intel versions.

I did start out by learning with uiSliders, and kept to very basic stuff while learning things in Ctrlr. It wasn't until I felt I had some footing that I moved on to my project. I hadn't tried to make a stand alone until I had something, and that's almost where I was when I started posting. I followed [spiffo's videos](https://youtu.be/SnQxErlcKIw) before I started making my panel.

It would seem that I need to chat with someone interested in moving the ARM Mac stuff forward. I don't know enough on my own to solve it, but I would be interested in helping.

**Reply by robucsc** [reply]
Oh and I forgot to say! The 5.3 version does update when knobs are turned on the synth! Which is way cool!

**Reply by damiensellier** [reply]
5.6 should do that too if midi i/o are working properly. Maybe it will be required to update CTRLR with the latest JUCE version to get it working on M1/M2 cpu. I only have a mac intel atm but I'll need a M2 soon so I'll work on this for sure.

**Reply by robucsc** [reply]
Sadly in 5.6 it doesn't work. 

I have started using system exclusive messages to the synth, and that is working well. When I send a request I can see the return message in the midi viewer, but that's as far as I've gotten. I don't think the panel sees it though.

**Reply by damiensellier** [reply]
No luck on this one. I guess using 5.3 in rosetta mode will be the only way until we dive into the dvpt of Apple silicon specific versions.

**Reply by robucsc** [reply]
Yup. I can put some effort into this, I'm just going to need a bit of help with getting it to compile, as I got stuck last time I tried it. Maybe we should make an ARM compile thread?

**Reply by damiensellier** [reply]
Yes it's a great idea. I'll do it right away to keep everything clean and gathered on this new thread.

**Reply by damiensellier** [reply]
Here we go : https://github.com/RomanKubiak/ctrlr/discussions/565

**Reply by robucsc** [reply]
Naisu! 

I'll give it a go tomorrow and see where I'm currently getting held up. 

Thanks!

**Reply by damiensellier** [reply]
If you can make an assessment of what's not working on the ARM64 version of CTRLR v5.6 with the exact same settings as 5.6 intel and 5.3.198 both with rosetta, that would be great.

**Comment by dnaldoog** [comment]
Hi @robusc,

With the fixed slider, set the values this way: I think the field is `Slider contents`.  I am not at a computer right now &#8211; The value before the = sign can be anything, even text.

<pre>
0=0
1=2
2=16
3=32
4=48
5=60
6=127
</pre>

**Reply by robucsc** [reply]
Cool thanks I will check this out!

**Reply by robucsc** [reply]
This worked fantastic!

The field is called uiFixedSliderContent

Thanks!
